include/core/checkpoint_control.h

```diff
@@ -0,0 +1,91 @@
+// SPDX-FileCopyrightText: Copyright (c) 2020-2024 The Lethe Authors
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception OR LGPL-2.1-or-later
+
+#ifndef lethe_checkpoint_control_h
+#define lethe_checkpoint_control_h
+
+#include <boost/archive/text_iarchive.hpp>
+#include <boost/archive/text_oarchive.hpp>
+
+class CheckpointControl
+{
+public:
+  /**
+   * @brief The CheckpointControl is responsible for keeping track of the
+   * next checkpoint id to use.
+   *
+   * @param checkpoint_freq Checkpoint frequency.
+   * @param restart_filename Filename used for the prefix during checkpointing.
+   */
+  CheckpointControl(const unsigned int checkpoint_freq,
+                    const std::string &restart_filename)
+    : max_checkpoint_id(2)
+    , next_checkpoint_id(max_checkpoint_id)
+    , checkpointing_frequency(checkpoint_freq)
+    , filename(restart_filename)
+  {}
+
+  /**
+   * @brief Return true if a checkpoint should be made at the current time step.
+   * @param time_step_number Current time step.
+   */
+  bool
+  is_checkpoint_time_step(const unsigned int time_step_number)
+  {
+    return (time_step_number % checkpointing_frequency) == 0;
+  }
+
+  /**
+   * @brief Calculate the next checkpoint id.
+   */
+  void
+  increment_checkpoint_id()
+  {
+    next_checkpoint_id = (next_checkpoint_id + 1) % max_checkpoint_id;
+  }
+
+  /**
+   * @brief Return the next checkpoint id.
+   */
+  unsigned int
+  get_next_checkpoint_id() const
+  {
+    return next_checkpoint_id;
+  }
+
+  /**
+   * @brief Return de prefix for the restart files.
+   */
+  std::string
+  get_filename() const
+  {
+    return filename;
+  }
+
+  /**
+   * @brief Serialize the checkpoint controller object to an output archive.
+   * @param ar Output archive where the attributes are stored.
+   */
+  void
+  serialize(boost::archive::text_oarchive &ar, const unsigned int) const
+  {
+    ar &next_checkpoint_id;
```

Comment:
I still find this notation (and the other below at line 82) confusing and I would prefer it to be << in this case ( and >> at line 82) for clarity.
In the end, it is up to you.

Reply:
I'M not sure the << and >> notation works for boost archives @PierreLaurentinCS . Everywhere in the code we used this notation.

Reply:
it works

Reply:
I was referring to the boost documentation (https://www.boost.org/doc/libs/1_39_0/libs/serialization/doc/archives.html) but it's true I did not test it. 
I think the notation is strictly equivalent. Though if & is used everywhere else in the code, then I see no reason not to use it here. It's not a big deal in the end.

Reply:
I'll let you decide what is best :)

Reply:
I'll keep this implementation to stay coherent with the rest of the code. When I'll fix the extra parameter in every serialize and deserialize function in the issue I have just opened, I will put the << and >> notation since I find in more intuitive. (the thing goes in or out of the archive.)